src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs

```diff
@@ -141,5 +130,97 @@ internal string GetMessageFromErrorResponse(HttpResponse response)
 
             return $"[Managed Identity] Error Code: {managedIdentityErrorResponse.Error} Error Message: {managedIdentityErrorResponse.ErrorDescription}";
         }
+
+        internal void HandleException(Exception ex, ManagedIdentitySource source = ManagedIdentitySource.None, string additionalInfo = null)
+        {
+            if (ex is HttpRequestException httpRequestException)
+            {
+                ThrowManagedIdentityException(
+                    MsalError.ManagedIdentityUnreachableNetwork,
+                    httpRequestException.Message,
+                    httpRequestException.InnerException);
+            }
+            else if (ex is TaskCanceledException taskCanceledException)
+            {
+                _requestContext.Logger.Error(TimeoutError);
+                throw taskCanceledException;
+            }
+            else if (ex is FormatException formatException)
+            {
+                string errorMessage = additionalInfo ?? formatException.Message;
+                _requestContext.Logger.Error($"[Managed Identity] Format Exception: {errorMessage}");
+                ThrowManagedIdentityException(
+                    MsalError.InvalidManagedIdentityEndpoint,
+                    errorMessage,
+                    formatException,
+                    source);
+            }
+            else if (ex is not MsalServiceException)
+            {
+                _requestContext.Logger.Error($"[Managed Identity] Exception: {ex.Message}");
+                
+                ThrowManagedIdentityException(
+                    MsalError.ManagedIdentityRequestFailed,
+                    ex.Message,
+                    ex);
+            }
+            else
+            {
+                // If it's already a MsalServiceException, rethrow it
+                throw ex;
+            }
+        }
+
+        internal void ThrowManagedIdentityException(
+            string errorCode,
+            string errorMessage,
+            Exception innerException = null,
+            ManagedIdentitySource managedIdentitySource = ManagedIdentitySource.None,
+            int? statusCode = null)
+        {
+            ManagedIdentitySource source = managedIdentitySource != ManagedIdentitySource.None ? managedIdentitySource : _sourceType;
+
+            ThrowServiceException(errorCode, errorMessage, innerException, source, statusCode);
+        }
+
+        internal static void ThrowServiceException(
+            string errorCode, 
+            string errorMessage,
+            Exception innerException,
+            ManagedIdentitySource managedIdentitySource,
+            int? statusCode = null)
```

Comment:
The factory has a better way of treating this. Why not just add something similar to https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/MsalServiceExceptionFactory.cs#L116